contracts/access/AccessControl.sol

```diff
@@ -0,0 +1,177 @@
+pragma solidity ^0.6.0;
+
+import "../utils/EnumerableSet.sol";
+
+/**
+ * @dev Contract module that allows children to implement role-based access
+ * control mechanisms.
+ *
+ * Roles are referred to by their `bytes32` identifier. These should be exposed
+ * in the external API and be unique. The best way to achieve this is by
+ * using `public constant` hash digests:
+ *
+ * ```
+ * bytes32 public constant MY_ROLE = keccak256("MY_ROLE");
+ * ```
+ *
+ * Roles can be used to represent a set of permissions. To restrict access to a
+ * function call, use {hasRole}:
+ *
+ * ```
+ * function foo() public {
+ *     require(hasRole(MY_ROLE, msg.sender));
+ *     ...
+ * }
+ * ```
+ *
+ * Roles can be granted and revoked programatically by calling the `internal`
+ * {_grantRole} and {_revokeRole} functions.
+ *
+ * This can also be achieved dynamically via the `external` {grantRole} and
+ * {revokeRole} functions. Each role has an associated admin role, and only
+ * accounts that have a role's admin role can call {grantRole} and {revokeRoke}.
+ *
+ * By default, the admin role for all roles is `DEFAULT_ADMIN_ROLE`, which means
+ * that only accounts with this role will be able to grant or revoke other
+ * roles. More complex role relationships can be created by using
+ * {_setRoleAdmin}.
+ */
+abstract contract AccessControl {
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    struct Role {
+        EnumerableSet.AddressSet members;
+        bytes32 admin; // This role's admin role
```

Comment:
Hm, good call. I refrained from adding the 'role' suffix because the struct itself is called `Role` (and the mapping `_roles`). I now think _those_ names might be off.

Still, as you pointed out, all of this is nonetheless internal.

Reply:
I dislike `admin`, but I think it might be `adminRoleId` instead. There's many mentions of `roleId`s, and a `Role` is the whole struct, not the `bytes32` identifier (the `roleId` :stuck_out_tongue:).

Still, as you pointed out, all of this is nonetheless internal.

Reply:
Either way is fine for me.